back/app/services/survey_results_generator_service.rb

```diff
@@ -107,70 +87,178 @@ def visit_file_upload(field)
     files = IdeaFile.where(id: file_ids).map do |file|
       { name: file.name, url: file.file.url }
     end
+    response_count = files.size
+    core_field_attributes(field, response_count).merge({
+      files: files
+    })
+  end
+
+  private
+
+  attr_reader :fields, :inputs, :locales
+
+  def core_field_attributes(field, response_count)
     {
       inputType: field.input_type,
       question: field.title_multiloc,
-      required: field.required,
-      totalResponses: files.size,
       customFieldId: field.id,
-      files: files
+      required: field.required,
+      grouped: false,
+      totalResponseCount: @inputs.count,
+      questionResponseCount: response_count
     }
   end
 
-  private
+  def visit_select_base(field, group_mode: nil, group_field_id: nil)
+    query = inputs
+    if group_field_id
+      if group_mode == 'user_field'
+        # Single user field grouped result
+        group_field = CustomField.find(group_field_id)
+        query = query.joins(:author)
+      else
+        # Single form field grouped result
+        group_field = find_question(group_field_id)
+      end
+      raise "Unsupported group field type: #{group_field.input_type}" unless group_field.input_type == 'select'
 
-  attr_reader :fields, :inputs, :locales
+      query = query.select(
+        select_field_query(field, as: 'answer'),
+        select_field_query(group_field, as: 'group')
+      )
+      answers = construct_grouped_answers(query, field, group_field)
+    else
+      query = query.select(
+        select_field_query(field, as: 'answer')
+      )
+      answers = construct_not_grouped_answers(query, field)
+    end
+
+    # Sort correctly
+    answers = answers.sort_by { |a| -a[:count] } unless field.input_type == 'linear_scale'
+    answers = answers.sort_by { |a| a[:answer] == 'other' ? 1 : 0 } # other should always be last
+
+    # Build response
+    build_select_response(answers, field, group_field)
+  end
+
+  def select_field_query(field, as: 'answer')
+    table = field.resource_type == 'User' ? 'users' : 'ideas'
 
-  def visit_select_base(field, values)
-    option_keys = field.options.pluck(:key)
-    distribution = Idea
-      .select(:value)
-      .from(values)
-      .group(:value)
-      .order(Arel.sql('COUNT(value) DESC'))
-      .count.to_a
-    (option_keys - distribution.pluck(0)).each do |key|
-      distribution << [key, 0] # add missing options with 0 responses
+    if %w[select linear_scale].include? field.input_type
+      "COALESCE(#{table}.custom_field_values->'#{field.key}', 'null') as #{as}"
+    else
+      %{
+          jsonb_array_elements(
+            CASE WHEN jsonb_path_exists(#{table}.custom_field_values, '$ ? (exists (@.#{field.key}))')
+              THEN #{table}.custom_field_values->'#{field.key}'
+              ELSE '[null]'::jsonb END
+          ) as #{as}
+      }
     end
```

Comment:
If I understand this piece of code correctly, would it make sense to make a bit clearer the condition when this code runs?
```suggestion
    elsif %w[multi_select].include? field.input_type
      %{
          jsonb_array_elements(
            CASE WHEN jsonb_path_exists(#{table}.custom_field_values, '$ ? (exists (@.#{field.key}))')
              THEN #{table}.custom_field_values->'#{field.key}'
              ELSE '[null]'::jsonb END
          ) as #{as}
      }
    else
      raise ArgumentError
    end
```
And the same can be done here https://github.com/CitizenLabDotCo/citizenlab/pull/7277/files?diff=split&w=1#diff-f519d3e628f000e1fe7fb51de96cb14194d9cefaa2e4bc308d781f53858c1419R119
But it's up to you really.